packages/affiliates/scripts/univ3.ts

```diff
@@ -0,0 +1,210 @@
+      multicall2Address: "0x5BA1e12693Dc8F9c48aAD8770482f4739bEeD696",
+      proxyAdminAddress: "0xd3808aBF85aC69D2DBf53794DEa08e75222Ad9a1",
+      tickLensAddress: "0x9dF511178D1438065F7672379414F5C46D5B51b4",
+      quoterAddress: "0x2051F6Fb61077b5A2A2c17535d31A1F2C858994f",
+      swapRouter: "0x58f6b77148BE49BF7898472268ae8f26377d0AA6",
+      nonfungibleTokenPositionDescriptorAddress: "0xeb86f5BE368c3C5e562f7eA1470ACC431d30fB0C",
+      descriptorProxyAddress: "0xB79bDE60fc227217f4EE2102dC93fa1264E33DaB",
+      nonfungibleTokenPositionManagerAddress: "0x865F20efC14A5186bF985aD42c64f5e71C055376",
+      v3MigratorAddress: "0x1988F2e49A72C4D73961C7f4Bb896819d3d2F6a3"
+    }
+  ]
+]);
+const infura = process.env.infura;
+const network = "rinkeby";
+
+function getAddress(network: NetworkName, contractName: string) {
+  const addresses = networks.get(network);
+  assert(addresses[contractName], "invalid contract name: " + contractName);
+  return addresses[contractName];
+}
+
+async function getPoolState(params: { pool: Pool; provider: any }) {
+  const { pool, provider } = params;
+  assert(pool.address, "requires pool address");
+  const contract = new ethers.Contract(pool.address, UniswapV3Pool.abi, provider);
+  const slot0 = await contract.slot0();
+  const liquidity = await contract.liquidity();
+  return {
+    ...slot0,
+    liquidity: liquidity.toString()
+  };
+}
+async function processNftEvents(params: { provider: any; positions: ReturnType<typeof NftPositions> }) {
+  const { provider, positions } = params;
+  const nftHandler = NftEvents({ positions });
+  const contract = new ethers.Contract(
+    getAddress(network, "nonfungibleTokenPositionManagerAddress"),
+    NFTPositionManager.abi,
+    provider
+  );
+  const events = await contract.queryFilter({});
+  await Promise.map(events, nftHandler);
+}
+async function getNftPositionState(params: { provider: any; position: NftPosition }) {
+  const { provider, position } = params;
+  const contract = new ethers.Contract(
+    getAddress(network, "nonfungibleTokenPositionManagerAddress"),
+    NFTPositionManager.abi,
+    provider
+  );
+  return convertValuesToString(await contract.positions(position.tokenId));
+}
+async function processPoolEvents(params: {
+  provider: any;
+  pool: Pool;
+  positions: ReturnType<typeof Positions>;
+  pools: ReturnType<typeof Pools>;
+}) {
+  const { pools, pool, provider, positions } = params;
+  assert(pool.id, "requires pool id");
+  assert(pool.address, "requires pool address");
+  const poolHandler = PoolEvents({ positions, id: pool.id, pools });
+  const contract = new ethers.Contract(pool.address, UniswapV3Pool.abi, provider);
+  const events = await contract.queryFilter({});
+  await Promise.map(events, poolHandler);
+}
+async function getPositionState(params: { position: Position; provider: any; pool: Pool }) {
+  const { position, provider, pool } = params;
+  assert(pool.address, "requires pool address");
```

Comment:
Same comment as above.

Reply:
ive taken a look at pool.address, and i am able to make that required without any trouble, but that's only because we are never doing any updates to the data in the pool table.  It would be awkward to have to force `{address}` to be included with any partial update to the pool. but since some fields are already being required now, im going to make it required.  will see how it goes.